### ISSUE TITLE
[Tests] Issues with signal delivery using tgkill() to threads

### ISSUE THREAD

**Issue by rsrinivas01** [lsds/sgx-lkl#518]
1. Explain what this test is doing, including a code snipped? We need to exact parameters when the system call is invoked ?

In LTP, There are total 3 files for tgkill related tests.

tgkill01.c:  
This test verifies that delivery of a signal to a specific thread is successful.  
This is done by 
 - installing a SIGUSR1 handler which records the current pthread ID.
 - number of threads are started in parallel, then  calling tgkill sys_tgkill(getpid(), threads[i].tid, SIGUSR1)
 - Check that the signal is delivered to the correct pthread. 

System call invocation:  TEST(sys_tgkill(getpid(), threads[i].tid, SIGUSR1));

tgkill02.c: 
This test verifies that tgkill() will fail with EAGAIN when RLIMIT_SIGPENDING is reached with a real-time signal.  This is done by
 - starting a child thread with SIGRTMIN blocked and a limit of 0 pending signals
 - attempting to deliver SIGRTMIN from the parent thread.

System call invocation: TEST(sys_tgkill(getpid(), tid, SIGRTMIN));

tgkill03.c:
This test verifies different tgkill() error cases 
This is done by invoking tgkill with different invalid values for tgid, tid and the signal parameters. 

System call invocation: TEST(sys_tgkill(*tc->tgid, *tc->tid, tc->sig)); with invalid values of tgid, tid and signal for different test cases. 

Issue with test failures:  The sys_tgkill( ) is returning error ESRCH (No process with the specified thread ID (and thread group ID) exists. This means either the thread or the thread group id doesn't exist. 

2. gdb and related stacktrace that shows what is causing the failure

For tgkill01:

[[  SGX-LKL ]] libc_start_main_stage2(): Calling app main: /ltp/testcases/kernel/syscalls/tgkill/tgkill01
tst_test.c:1106: INFO: Timeout per run is 0h 05m 00s
tst_test.c:1125: INFO: No fork support
tst_checkpoint.c:149: BROK: tgkill01.c:47: tst_checkpoint_wait(1, 10000): ??? (516)
tgkill01.c:94: FAIL: SIGUSR1 delivered to wrong thread
tst_checkpoint.c:162: WARN: tgkill01.c:63: tst_checkpoint_wake(1, 10, 10000): ETIMEDOUT (110)
tst_checkpoint.c:162: WARN: tgkill01.c:63: tst_checkpoint_wake(1, 10, 10000): ETIMEDOUT (110)
tgkill01.c:69: WARN: pthread_join(..., 0) failed: EINVAL
[[  SGX-LKL ]] lkl_terminate(): terminating LKL (exit_status=0)
[[  SGX-LKL ]] lkl_termination_thread(): termination thread unblocked
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_chdir(/)
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_umount_timeout(/)
[   13.045902] EXT4-fs (vda): re-mounted. Opts: (null)
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_virtio_netdev_remove()
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_halt()
[   13.046459] reboot: Restarting system
[[  SGX-LKL ]] lkl_termination_thread(): lthread_detach2() done
[   SGX-LKL  ] ethread (5: 0) ethread (7: 0) ethread (6: 0) ethread (3: 0) ethread (4: 0) ethread (1: 0) [Thread 0x7fff89ffb700 (LWP 20221) exited]
[Thread 0x7fff8a7fc700 (LWP 20220) exited]
[Thread 0x7fff8affd700 (LWP 20219) exited]
[Thread 0x7fff8b7fe700 (LWP 20218) exited]
[Thread 0x7fff90f9d700 (LWP 20217) exited]
[Thread 0x7fff8bfff700 (LWP 20215) exited]
init (0: 0 exit=0) ethread (2: 0) [Thread 0x7fff9179e700 (LWP 20216) exited]
[Thread 0x7fff91f9f700 (LWP 20214) exited]

[   SGX-LKL  ] oe_terminate_enclave... done
[   SGX-LKL  ] SGX-LKL-OE exit: exit_status=0
[Thread 0x7fff92fa1700 (LWP 20212) exited]
[Thread 0x7fff937a2700 (LWP 20211) exited]
[Thread 0x7fff93fa3700 (LWP 20210) exited]
[Thread 0x7ffff7fa4680 (LWP 20200) exited]
[Inferior 1 (process 20200) exited normally]
(gdb)

For tgkill02:

[[  SGX-LKL ]] libc_start_main_stage2(): Calling app main: /ltp/testcases/kernel/syscalls/tgkill/tgkill02
tst_test.c:1106: INFO: Timeout per run is 0h 05m 00s
tst_test.c:1125: INFO: No fork support
tgkill02.c:58: FAIL: tgkill() should have failed with EAGAIN: ESRCH (3)
[[  SGX-LKL ]] lkl_terminate(): terminating LKL (exit_status=0)
[[  SGX-LKL ]] lkl_termination_thread(): termination thread unblocked
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_chdir(/)
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_umount_timeout(/)
[    0.397888] EXT4-fs (vda): re-mounted. Opts: (null)
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_virtio_netdev_remove()
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_halt()
[    0.398445] reboot: Restarting system
[[  SGX-LKL ]] lkl_termination_thread(): lthread_detach2() done
[   SGX-LKL  ] ethread (3: 0) init (0: 0 exit=0) ethread (5: 0) ethread (7: 0) ethread (4: 0) ethread (1: 0) ethread (2: 0) [Thread 0x7fff89ffb700 (LWP 20665) exited]
ethread (6: 0) [Thread 0x7fff8affd700 (LWP 20663) exited]
[Thread 0x7fff8a7fc700 (LWP 20664) exited]

[   SGX-LKL  ] [Thread 0x7fff8b7fe700 (LWP 20662) exited]
oe_terminate_enclave... [Thread 0x7fff8bfff700 (LWP 20661) exited]
[Thread 0x7fff90f9d700 (LWP 20660) exited]
[Thread 0x7fff9179e700 (LWP 20659) exited]
[Thread 0x7fff91f9f700 (LWP 20658) exited]
done
[   SGX-LKL  ] SGX-LKL-OE exit: exit_status=0
[Thread 0x7fff92fa1700 (LWP 20656) exited]
[Thread 0x7fff937a2700 (LWP 20655) exited]
[Thread 0x7fff93fa3700 (LWP 20654) exited]
[Thread 0x7ffff7fa4680 (LWP 20640) exited]
[Inferior 1 (process 20640) exited normally]
(gdb)

For tgkill03:

[[  SGX-LKL ]] libc_start_main_stage2(): Calling app main: /ltp/testcases/kernel/syscalls/tgkill/tgkill03
tst_test.c:1106: INFO: Timeout per run is 0h 00m 20s
tst_test.c:1125: INFO: No fork support
tgkill03.c:95: PASS: Invalid tgid failed as expected: EINVAL (22)
tgkill03.c:95: PASS: Invalid tid failed as expected: EINVAL (22)
tgkill03.c:99: FAIL: Invalid signal should have failed with EINVAL: ESRCH (3)
tgkill03.c:95: PASS: Defunct tid failed as expected: ESRCH (3)
tgkill03.c:95: PASS: Defunct tgid failed as expected: ESRCH (3)
tgkill03.c:104: FAIL: Valid tgkill call failed: ESRCH (3)
[[  SGX-LKL ]] lkl_terminate(): terminating LKL (exit_status=0)
[[  SGX-LKL ]] lkl_termination_thread(): termination thread unblocked
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_chdir(/)
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_umount_timeout(/)
[    0.407592] EXT4-fs (vda): re-mounted. Opts: (null)
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_virtio_netdev_remove()
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_halt()
[    0.407592] reboot: Restarting system
[[  SGX-LKL ]] lkl_termination_thread(): lthread_detach2() done
[   SGX-LKL  ] init (0: 0 exit=0) ethread (3: 0) ethread (4: 0) ethread (1: 0) ethread (5: 0) ethread (2: 0) ethread (7: 0) ethread (6: 0) [Thread 0x7fff82ffd700 (LWP 20751) exited]
[Thread 0x7fff81ffb700 (LWP 20753) exited]

[Thread 0x7fff827fc700 (LWP 20752) exited]
[   SGX-LKL  ] oe_terminate_enclave... [Thread 0x7fff837fe700 (LWP 20750) exited]
[Thread 0x7fff83fff700 (LWP 20749) exited]
[Thread 0x7fff90f9d700 (LWP 20748) exited]
[Thread 0x7fff9179e700 (LWP 20747) exited]
[Thread 0x7fff91f9f700 (LWP 20746) exited]
done
[   SGX-LKL  ] SGX-LKL-OE exit: exit_status=0
[Thread 0x7fff927a0700 (LWP 20745) exited]
[Thread 0x7fff92fa1700 (LWP 20744) exited]
[Thread 0x7fff93fa3700 (LWP 20742) exited]
[Thread 0x7ffff7fa4680 (LWP 20720) exited]
[Inferior 1 (process 20720) exited normally]
(gdb)

**Comment by rsrinivas01** [comment]
@prp @hukoyu I have updated the issue to include more technical information and corresponding gdb traces.

**Comment by prp** [comment]
Thanks. Can you also use proper markdown formatting in the initial comment?

Some of the documented errors should be addressed by the relayering work (e.g. https://github.com/lsds/sgx-lkl/issues/182). For now, this test will be broken and should be deactivated.

**Comment by SeanTAllen** [comment]
@prp @davidchisnall is this a duplicate of #644? To me, it feels like this could be closed and added as a note to #644 to reenable the test.

**Comment by SeanTAllen** [comment]
Closing in favor of #644. 

I added a note about enabling the tgkill tests as part of work on 644.
